storage/sql.go

```diff
@@ -289,3 +296,165 @@ func verifyModel(ctx context.Context, db *pg.DB, m *orm.Table) error {
 func stripQuotes(s types.Safe) string {
 	return strings.Trim(string(s), `"`)
 }
+
+func (d *Database) LeaseBlocks(ctx context.Context, claimUntil time.Time, batchSize int) (visor.ProcessingBlockList, error) {
+	var blocks visor.ProcessingBlockList
+
+	if err := d.DB.RunInTransaction(ctx, func(tx *pg.Tx) error {
+		_, err := tx.QueryContext(ctx, &blocks, `
+WITH leased AS (
+    UPDATE visor_processing_blocks
+    SET claimed_until = ?
+    FROM (
+	    SELECT *
+	    FROM visor_processing_blocks
+	    WHERE completed_at IS null AND
+	          (claimed_until IS null OR claimed_until < ?) AND
+	          height > 0
+	    ORDER BY height DESC
+	    LIMIT ?
```

Comment:
FYI the prior selection algorithm used `RANK()` to ensure we were making batch selections which fall on tipset boundaries (such that some blocks from a tipset are not divided across two processing batches). This would cause problems with the state resolution if we separated blocks.

Reply:
Noted. I'll adjust.

Reply:
I think this method should return a list of tipsets

Reply:
Reworked actor stage change detection and message processing to use tipsets instead of blocks. Renamed the `visor_processing_blocks` table to `visor_processing_statechanges` since it's a record of which tipsets have been analysed for actor state changes. It now has a tipset column instead of cid.